### PR TITLE
QGIS: 2.18.17 -> 3.0.1 plus required libs

### PR DESCRIPTION
**Pull request by coreyoconnor**
I think the wrapper could use some improvements, but it's functional.

###### Motivation for this change

1. update QGIS to 3.0.1

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -10065,6 +10065,27 @@ in {
     };
   };
 
+  OWSLib = buildPythonPackage rec {
```
Comment (jtojnar) [inline pkgs/top-level/python-packages.nix]: Could you move this to a separate file?

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -10065,6 +10065,27 @@ in {
...
+    pname = "OWSLib";
+    version = "0.16.0";
+    name = "${pname}-${version}";
+
+    src = self.fetchPypi {
+      inherit pname version;
+      sha256 = "0m05225g1sqd2i8r2riaan33953hfni9wjq7n225snhl7klsb5gc";
+    };
+
+    buildInputs = with self; [ dateutil pep8 pillow pyproj pytz pytest pytestcov requests tox ];
```
Comment (jtojnar) [inline pkgs/top-level/python-packages.nix]: Some of these should probably go to `checkInputs`.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -10065,6 +10065,27 @@ in {
...
+
+    doCheck = false;
+
+    meta = {
+      description = "Client programming with Open Geospatial Consortium (OGC) web service.";
+      license = licenses.bsd3;
+      homepage = "http://geopython.github.io/OWSLib/";
```
Comment (jtojnar) [inline pkgs/top-level/python-packages.nix]: URLs need not be quoted.

Comment (jtojnar) [inline pkgs/top-level/python-packages.nix]: URLs do not need to be quoted.

File: pkgs/top-level/qscintilla-pyqt5-widgets.patch [hunk 4]
```diff
@@ -0,0 +1,12 @@
+diff -u -r orig/Python/configure.py QScintilla_gpl-2.10.3/Python/configure.py
```
Comment (jtojnar) [inline pkgs/top-level/qscintilla-pyqt5-widgets.patch]: This definitely cannot be in top-level. Should go to `pkgs/development/python-modules/qscintilla/`.

File: pkgs/applications/gis/qgis/default.nix [hunk 5]
```diff
@@ -1,42 +1,48 @@
...
-  src = fetchurl {
-    url = "http://qgis.org/downloads/${name}.tar.bz2";
-    sha256 = "1nxwl5lwibbiz9v3qaw3px7iyxg113zr4j8d99yj07mhk2ap082y";
-  };
-
-  cmakeFlags = stdenv.lib.optional withGrass "-DGRASS_PREFIX7=${grass}/${grass.name}";
+  # Build processes depend on the built libraries. Extend
+  # LD_LIBRARY_PATH as required.
+  preConfigure = ''
+    export LD_LIBRARY_PATH=`pwd`/build/output/lib:${stdenv.lib.makeLibraryPath [ openssl ]}$LD_LIBRARY_PATH
```
Comment (jtojnar) [inline pkgs/applications/gis/qgis/default.nix]: Would not adding `-DCMAKE_SKIP_BUILD_RPATH=OFF` to `cmakeFlags` remove the need for this hack this?

File: pkgs/applications/gis/qgis/default.nix [hunk 6]
```diff
@@ -1,42 +1,48 @@
...
+    fcgi libspatialindex libspatialite postgresql qjson txt2tags libzip
+    qtbase qtwebkit qtsensors qca-qt5 qtkeychain qscintilla ] ++
+    (stdenv.lib.optional withGrass grass) ++ pythonBuildInputs;
 
-  # To handle the lack of 'local' RPATH; required, as they call one of
-  # their built binaries requiring their libs, in the build process.
-  preBuild = ''
-    export LD_LIBRARY_PATH=`pwd`/output/lib:${stdenv.lib.makeLibraryPath [ openssl ]}:$LD_LIBRARY_PATH
-  '';
+  nativeBuildInputs = [ cmake ];
```
Comment (jtojnar) [inline pkgs/applications/gis/qgis/default.nix]: Please add `ninja` when using `cmake`, it is usually faster than make.

File: pkgs/top-level/qscintilla-pyqt5-widgets.patch (hunk 4 above)
Comment (coreyoconnor) [reply pkgs/top-level/qscintilla-pyqt5-widgets.patch]: whoops! editor goof

File: pkgs/applications/gis/qgis/default.nix (hunk 5 above)
Comment (coreyoconnor) [reply pkgs/applications/gis/qgis/default.nix]: No idea. That was from the previous QGIS version. I'll try that option and see.

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (coreyoconnor) [reply pkgs/top-level/python-packages.nix]: Yes. There is a mix of inline and separate file in python-packages. Is the standard to move to separate file?

Comment (jtojnar) [reply pkgs/top-level/python-packages.nix]: In the past having module expressions in this file was okay; nowadays, we are moving them to separate files.

File: pkgs/applications/gis/qgis/default.nix [hunk 7]
```diff
@@ -1,42 +1,48 @@
...
-    fcgi libspatialindex libspatialite postgresql qjson qca2 txt2tags ] ++
-    (stdenv.lib.optional withGrass grass) ++
-    (with python2Packages; [ jinja2 numpy psycopg2 pygments requests python2Packages.qscintilla sip ]);
+  src = fetchurl {
+    url = "http://qgis.org/downloads/qgis-${version}.tar.bz2";
+    sha256 = "1m24kjl784csbv0dgx1wbdwg8r92cpc1j718aaw85p7vgicm8acy";
+  };
 
-  nativeBuildInputs = [ cmake makeWrapper ];
+  pythonBuildInputs = [ python3Packages.qscintilla python3Packages.gdal ] ++
```
Comment (FRidh) [inline pkgs/applications/gis/qgis/default.nix]: This should go into ` let ... in` expression.

File: pkgs/applications/gis/qgis/wrapped.nix [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  postBuild = ''
+    unpackPhase
+
+    buildPythonPath "$pythonInputs"
+
+    wrapProgram $out/bin/qgis \
+      --prefix PATH : $program_PATH \
+      --prefix PYTHONPATH : $program_PYTHONPATH \
+      --prefix LD_LIBRARY_PATH : ${stdenv.lib.makeLibraryPath [ openssl ]}
```
Comment (FRidh) [inline pkgs/applications/gis/qgis/wrapped.nix]: this should be patched instead.

File: pkgs/applications/gis/qgis/wrapped.nix [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  inherit (qgis-unwrapped) src;
+
+  postBuild = ''
+    unpackPhase
+
+    buildPythonPath "$pythonInputs"
+
+    wrapProgram $out/bin/qgis \
+      --prefix PATH : $program_PATH \
+      --prefix PYTHONPATH : $program_PYTHONPATH \
```
Comment (FRidh) [inline pkgs/applications/gis/qgis/wrapped.nix]: No prefixing, but setting so PYTHONPATH cannot leak inside.

File: pkgs/top-level/all-packages.nix [hunk 10]
```diff
@@ -17335,7 +17335,11 @@ with pkgs;
 
   qemu-riscv = callPackage ../applications/virtualization/qemu/riscv.nix {};
 
-  qgis = callPackage ../applications/gis/qgis {};
+  qgis-unwrapped = libsForQt5.callPackage ../applications/gis/qgis {
```
Comment (FRidh) [inline pkgs/top-level/all-packages.nix]: Why expose the unwrapped version? Is it used somewhere?

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (FRidh) [reply pkgs/top-level/python-packages.nix]: We are moving Python expressions out of `pkgs/top-level/python-packages.nix` into `pkgs/development/python-modules/<module>/default.nix`. 

Please move the expression there, and call it from `pkgs/top-level/python-packages.nix` using `callPackage ../development/python-modules/<package> { };`.

File: pkgs/top-level/python-packages.nix [hunk 11]
```diff
@@ -10065,6 +10065,27 @@ in {
     };
   };
 
+  OWSLib = buildPythonPackage rec {
+    pname = "OWSLib";
+    version = "0.16.0";
+    name = "${pname}-${version}";
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: The `name` attribute is added by `buildPython*` and should therefore be removed.

**Review by FRidh — CHANGES_REQUESTED** [review]
We should not yet replace 2.8 by 3.0 because 3.0 is not yet feature complete. Instead, add 3.0 next to 2.8.

File: pkgs/applications/gis/qgis/default.nix (hunk 5 above)
Comment (jtojnar) [reply pkgs/applications/gis/qgis/default.nix]: Apparently the `preConfigure` section is not required any more.

File: pkgs/applications/gis/qgis/wrapped.nix (hunk 8 above)
Comment (jtojnar) [reply pkgs/applications/gis/qgis/wrapped.nix]: This is probably not necessary anymore. `ldd $(nix-build -A qgis)/bin/.qgis-wrapped | grep openssl` shows it is linked correctly.

**Comment by jtojnar** [comment]
> We should not yet replace 2.8 by 3.0 because 3.0 is not yet feature complete.

Do you have a source for that? I only found

> This is the first release in the 3.x series. It comes with tons of new features and under-the-hood updates. As such, we do not expect it to be as reliable as the 2.18 LTR just yet.

https://qgis.org/en/site/forusers/visualchangelog30/index.html

**Comment by FRidh** [comment]
Its actually the plugins I meant to refer to. Plugins written in 2 are not compatible with 3. As they are a major feature of QGIS we should not be dropping 2 yet. Before 18.09 we can revisit this choice.

File: pkgs/top-level/all-packages.nix (hunk 10 above)
Comment (coreyoconnor) [reply pkgs/top-level/all-packages.nix]: Exposing the unwrapped version in these cases is easiest way (IMO) to enable an overlay changing the source of the application. EG:

* https://github.com/coreyoconnor/nix_configs/commit/c37c85afa312e8ba79f9b96cbd4f49add62a9b56

**Comment by coreyoconnor** [comment]
That explains the mixed messaging I've seen online. The version number and some messaging indicates the QGIS 3 is released. Other messaging (EG: loading screen) states "early adopter".

I just figured it was inconsistency haha. The plugin situation explains that. I'll re-add the 2 derivation.

**Comment by coreyoconnor** [comment]
I re-added QGIS 2.18 as well as resolved the noted issues. Other changes with latest commits:

* Qscintilla is 2.9 not 2.10 as in the original commits. 2.10 has a different library naming convention (`_qtN` suffix) which would change the interface for any existing qscintilla dependents. I don't know if that would matter so I'll defer the upgrade till later. 
* partitioned the python package changes into separate commits.

File: pkgs/applications/gis/qgis/3.0.nix [hunk 12]
```diff
@@ -0,0 +1,50 @@
+{ stdenv, lib, fetchurl, cmake, ninja, flex, bison, proj, geos, xlibsWrapper, sqlite, gsl
+, qwt, fcgi, python3Packages, libspatialindex, libspatialite, postgresql
+, qjson, txt2tags, openssl, libzip
+, qtbase, qtwebkit, qtsensors, qca-qt5, qtkeychain, qscintilla
+, withGrass ? true, grass
+}:
+with lib;
+let
+  pythonBuildInputs = with python3Packages;
+    [ python3Packages.qscintilla gdal jinja2 numpy psycopg2 pygments pyqt5 sip OWSLib six ];
```
Comment (FRidh) [inline pkgs/applications/gis/qgis/3.0.nix]: repeating `python3Packages`

File: pkgs/development/python-modules/OWSLib/default.nix [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0m05225g1sqd2i8r2riaan33953hfni9wjq7n225snhl7klsb5gc";
+  };
+
+  buildInputs = [ dateutil pep8 pillow pyproj pytz requests tox ];
+  checkInputs = [ pytest pytestcov ];
+
+  doCheck = false;
```
Comment (FRidh) [inline pkgs/development/python-modules/OWSLib/default.nix]: Why are the tests disabled? Include a comment in the expression explaining why the tests are disabled.

File: pkgs/development/python-modules/OWSLib/default.nix [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+buildPythonPackage rec {
+  pname = "OWSLib";
+  version = "0.16.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0m05225g1sqd2i8r2riaan33953hfni9wjq7n225snhl7klsb5gc";
+  };
+
+  buildInputs = [ dateutil pep8 pillow pyproj pytz requests tox ];
```
Comment (FRidh) [inline pkgs/development/python-modules/OWSLib/default.nix]: `cehckInputs`?

**Review by FRidh — CHANGES_REQUESTED** [review]
Why is the wrapper needed? Can't this be done in the `installPhase` or `fixupPhase`?

Of course it is more convenient when designing the wrappers when not needing to rebuild the whole thing.

File: pkgs/applications/gis/qgis/3.0.nix (hunk 12 above)
Comment (coreyoconnor) [reply pkgs/applications/gis/qgis/3.0.nix]: Not a repeat. The "with" does not cause the "qscintilla" to be shadowed. Without the extra python3Packages the "qscintilla" passed is not the one in python3Packages but the C++ library one.

File: pkgs/development/python-modules/OWSLib/default.nix (hunk 13 above)
Comment (coreyoconnor) [reply pkgs/development/python-modules/OWSLib/default.nix]: No tests run and the build fails because no test run. I figured this was a bug in the python build infrastructure.

**Comment by coreyoconnor** [comment]
The wrapper is required to enable extension of the qgis python environment by users without requiring a rebuild of the full application. EG: to add additional deps for plugins for instance.

File: pkgs/development/python-modules/OWSLib/default.nix (hunk 14 above)
Comment (coreyoconnor) [reply pkgs/development/python-modules/OWSLib/default.nix]: Already resolved. Not sure why this diff is showing up for you.

File: pkgs/applications/gis/qgis/wrapped.nix [hunk 15]
```diff
@@ -0,0 +1,46 @@
+{ stdenv, lib, fetchurl, openssl, python3Packages, makeWrapper, symlinkJoin
+, qgis3-unwrapped
+}:
+with lib;
+let
+  # the python packages that should be included in the PYTHONPATH
+  pythonInputs = with python3Packages;
+    qgis3-unwrapped.pythonBuildInputs ++ [ chardet dateutil pyyaml pytz requests urllib3 ];
```
Comment (FRidh) [inline pkgs/applications/gis/qgis/wrapped.nix]: Why should these few be included in the wrapper, but *not* in the build? Are they needed for specific plugins?

Comment (coreyoconnor) [reply pkgs/applications/gis/qgis/wrapped.nix]: Moved to build. The reference errors these resolved only showed up at runtime.
I would expect these are transitive dependencies for packages missing respective propagatedBuildInputs?

**Comment by coreyoconnor** [comment]
Amended to resolve current CRs.

File: pkgs/applications/gis/qgis/wrapped.nix (hunk 15 above)
Comment (FRidh) [reply pkgs/applications/gis/qgis/wrapped.nix]: They are not hard dependencies otherwise the builds of the dependents would fail. Either upstream has forgotten to list their full dependencies, or they require "extras" of certain packages.
https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

File: pkgs/applications/gis/qgis/wrapped.nix [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+  postBuild = ''
+    unpackPhase
+
+    buildPythonPath "$pythonInputs"
+
+    wrapProgram $out/bin/qgis \
+      --prefix PATH : $program_PATH \
+      --set PYTHONPATH $program_PYTHONPATH
+
+    # desktop link
```
Comment (FRidh) [inline pkgs/applications/gis/qgis/wrapped.nix]: The following seems irrelevant for the wrapper and thus better placed in the original derivation. Or am I missing something?

**Comment by coreyoconnor** [comment]
Times up! no further changes will be submitted.

**Comment by mpickering** [comment]
This looks like a great start. I was blocked on this as some dependencies were not building for darwin. If they are resolved I can pick it up.

**Comment by das-g** [comment]
I successfully build QGIS 3 from this branch on NixOS 18.03.

How do we proceed? Shall I attempt to resolve the merge conflicts with `master`?

> Times up! no further changes will be submitted.

What does that mean? Was your effort on this time-boxed @coreyoconnor or are you referring to NixOS release 18.03?

**Comment by das-g** [comment]
QGIS 3.2 Bonn [has been released](http://blog.qgis.org/2018/06/24/qgis-3-2-bonn-is-released/). Time to revive this PR?

**Comment by coreyoconnor** [comment]
Hi! The time I had for this original PR ran out a while ago. I will not have a chance for a few weeks (sadly) to really dig into this again. Anyone should feel free to take the changes thus far and carry on, of course.

I do have a separate branch with these changes rebased against HEAD. Might be able to get this updated accordingly but besides that... a few weeks it is.

**Comment by coreyoconnor** [comment]
Clearing old PRs.

**Comment by das-g** [comment]
New PR for QGIS 3.4.5 (=the new LTR version and now currently only LTR version): #56486
